### PR TITLE
feat: optimize component logic | authorization | services | hooks

### PR DESCRIPTION
**Pull request by VitaliyOl**

### CODE REVIEW

File: packages/frontend/src/Pages/HomePage.tsx [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    return <TodosPage />;
+  }
+
+  return (
+    <div>
+      <h1>Login</h1>     
+      <LoginForm />
+      <p>
+        If you are not registered, please{' '}
+        <Link to="/register">register here</Link>.
```
Comment (malyshkoserhii) [inline packages/frontend/src/Pages/HomePage.tsx:20]: ROUTER_KEYS

File: packages/frontend/src/components/Auth/LoginForm/LoginForm.tsx [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  const location = useLocation();
+
+  const { formik, errorMessage, setErrorMessage } = useCustomFormik<LoginFormValues>({
+    initialValues: loginInitialValues,
+    validationSchema: loginSchema,
+    onSubmit: async (values) => {
+      try {        
+        const response = await http.login<LoginFormValues>('user/login', values);
+        localStorage.setItem(STORAGE_KEYS.USER_ID, response.data.user.id.toString());
+        navigate('/todos');
```
Comment (malyshkoserhii) [inline packages/frontend/src/components/Auth/LoginForm/LoginForm.tsx:25]: ROUTER_KEYS

File: packages/frontend/src/components/Auth/LoginForm/LoginForm.tsx [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import { loginInitialValues } from '~/shared/services/initialValues';
+
+const LoginForm: React.FC = () => {
+  const navigate = useNavigate();
+  const location = useLocation();
+
+  const { formik, errorMessage, setErrorMessage } = useCustomFormik<LoginFormValues>({
+    initialValues: loginInitialValues,
+    validationSchema: loginSchema,
+    onSubmit: async (values) => {
```
Comment (malyshkoserhii) [inline packages/frontend/src/components/Auth/LoginForm/LoginForm.tsx:21]: avoid inline fucntions

File: packages/frontend/src/components/Auth/LoginForm/LoginForm.tsx [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+      try {        
+        const response = await http.login<LoginFormValues>('user/login', values);
+        localStorage.setItem(STORAGE_KEYS.USER_ID, response.data.user.id.toString());
+        navigate('/todos');
+      } catch (error: any) {
```
Comment (malyshkoserhii) [inline packages/frontend/src/components/Auth/LoginForm/LoginForm.tsx:26]: any

File: packages/frontend/src/components/Auth/LoginForm/LoginForm.tsx [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        if (error.response && error.response.status === 401) {
+          setErrorMessage('Invalid email or password. Please try again.');
```
Comment (malyshkoserhii) [inline packages/frontend/src/components/Auth/LoginForm/LoginForm.tsx:28]: send error message from your backend

File: packages/frontend/src/components/Auth/LoginForm/LoginForm.tsx [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  });
+
+  useEffect(() => {
+    if (location.state && location.state.registered) {
+      setErrorMessage('Verification email sent. Please check your email to verify your account.');
+    }
+  }, [location, setErrorMessage]);
+
+  return (
+    <Form id="login-form" onSubmit={formik.handleSubmit} className={formStyles}>
```
Comment (malyshkoserhii) [inline packages/frontend/src/components/Auth/LoginForm/LoginForm.tsx:43]: add validation for your form

File: packages/frontend/src/components/Auth/RegisterForm/RegisterForm.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+const RegisterForm: React.FC = () => {
+  const navigate = useNavigate();
+
+  const { formik, errorMessage } = useCustomFormik<RegisterFormValues>({
+    initialValues: registerInitialValues,
+    validationSchema: registerSchema,
+    onSubmit: async (values) => {      
+      await http.post('user/register', values);
+      navigate('/login', { state: { registered: true } });
```
Comment (malyshkoserhii) [inline packages/frontend/src/components/Auth/RegisterForm/RegisterForm.tsx:22]: ROUTER_KEYS

File: packages/frontend/src/hooks/useCustomFormik.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+export const useCustomFormik = <Values extends Record<string, any>>(
+  config: FormikConfig<Values>
+): UseCustomFormikReturn<Values> => {
+  const [errorMessage, setErrorMessage] = useState<string | null>(null);
+
+  const onSubmit = async (values: Values, formikHelpers: FormikHelpers<Values>) => {
+    try {
+      await config.onSubmit(values, formikHelpers);
+      setErrorMessage(null);
+    } catch (error: any) {
```
Comment (malyshkoserhii) [inline packages/frontend/src/hooks/useCustomFormik.ts:19]: any

File: packages/frontend/src/router/ProtectedRoute.tsx [hunk 9]
```diff
@@ -0,0 +1,19 @@
+import React from 'react';
+import { Navigate, Outlet } from 'react-router-dom';
+import { STORAGE_KEYS } from '~shared/keys';
+
+interface ProtectedRouteProps {
+  redirectPath?: string;
+}
+
+const ProtectedRoute: React.FC<ProtectedRouteProps> = ({ redirectPath = '/login' }) => {
+  const token = localStorage.getItem(STORAGE_KEYS.TOKEN);
```
Comment (malyshkoserhii) [inline packages/frontend/src/router/ProtectedRoute.tsx:10]: use Zustand persist approach to define is user logged in or not

File: packages/frontend/src/shared/services/http.ts [hunk 10]
```diff
@@ -1,75 +1,77 @@
...
+      delete this.fetchingService.defaults.headers.common['Authorization'];
+    }
+  }
+  
+  setToken(token: string): void {
+    localStorage.setItem(STORAGE_KEYS.TOKEN, token);
+    this.setAuthorizationHeader();
+  }
+
+  private handleRequest<T>(method: 'get' | 'post' | 'put' | 'delete', url: string, data?: any, config: AxiosRequestConfig = {}): Promise<AxiosResponse<T>> {
```
Comment (malyshkoserhii) [inline packages/frontend/src/shared/services/http.ts:33]: createe separate type for method

**Comment by VitaliyOl** [comment]
I have fixed! thanks

**Pull request merged**